File: src/main/java/calclavia/lib/grid/Grid.java
package calclavia.lib.grid;

import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import java.util.WeakHashMap;

/**
 * A grid specifying a connection with a series of nodes.
 * 
 * @author Calclavia
 * 
 * @param <N> - The node type.
 */
public abstract class Grid<N>
{
	/**
	 * A set of connectors (e.g conductors).
	 */
	protected final Set<N> nodes = Collections.newSetFromMap(new WeakHashMap<N, Boolean>());
	private final Class<? extends N> nodeType;

	public Grid(Class<? extends N> type)
	{
		nodeType = type;
	}

	public void add(N node)
	{
		synchronized (nodes)
		{
			nodes.add(node);
		}
	}

	public void remove(N node)
	{
		synchronized (nodes)
		{
			nodes.remove(node);
		}
	}

	public Set<N> getNodes()
	{
		return nodes;
	}

	/**
	 * A simple reconstruct class to rebuild the grid. The set "nodes" is copied due to the fact
	 * that this method will allow the modification of nodes while looping.
	 */
	public void reconstruct()
	{
		synchronized (nodes)
		{
			Iterator<N> it = new HashSet<N>(nodes).iterator();

			while (it.hasNext())
			{
				N node = it.next();

				if (isValidNode(node))
				{
					reconstructNode(node);
				}
				else
				{
					it.remove();
				}
			}
		}
	}

	public boolean isValidNode(Object node)
	{
		return nodeType.isAssignableFrom(node.getClass());
	}

	protected void reconstructNode(N node)
	{

	}

	public void deconstruct()
	{
		synchronized (nodes)
		{
			nodes.clear();
		}
	}

	/**
	 * Gets the first connector in the set.
	 * 
	 * @return
	 */
	public N getFirstNode()
	{
		synchronized (nodes)
		{
			for (N node : nodes)
			{
				return node;
			}
		}
		return null;
	}

	@Override
	public String toString()
	{
		return getClass().getSimpleName() + "[" + hashCode() + ", Nodes: " + nodes.size() + "]";
	}
}

File: src/main/scala/universalelectricity/core/grid/Grid.scala
package universalelectricity.core.grid

import java.util
import java.util.Collections

import scala.collection.convert.wrapAll._
import scala.reflect.ClassTag

/**
 * A grid containing a series of arbitrary nodes.
 *
 * @author Calclavia
 */
class Grid[N: ClassTag]
{
  /** A set of nodes in the grid (e.g conductors). */
  private final val nodes = Collections.newSetFromMap(new util.WeakHashMap[N, java.lang.Boolean]())

  /** A class instance of the node type used to node validation checks. */
  val nodeClass = implicitly[ClassTag[N]].runtimeClass

  /**
   * Adds a node to the grid.
   * @param node
   */
  def add(node: N)
  {
    nodes.add(node)
  }

  /**
   * Removes a node to the grid.
   * @param node
   */
  def remove(node: N)
  {
    nodes.remove(node)
  }

  /**
   * Make sure nodes are synchronized if modification is required.
   *
   * @return The list of nodes.
   */
  def getNodes() = nodes

  /**
   * Rebuilds the grid. The set "nodes" is copied due to the fact that this method will allow the modification of nodes while looping on a separate thread.
   */
  def reconstruct()
  {
    nodes.filter(isValidNode(_)).foreach(reconstructNode(_))
  }

  def isValidNode(node: Any): Boolean =
  {
    //TODO: Check if better way to do this.
    return nodeClass.isAssignableFrom(node.getClass)
   }

  protected def reconstructNode(node: N)
  {
  }

  /**
   * Clears all nodes
   */
  def deconstruct()
  {
    nodes.clear()
  }

  /**
   * Gets the first node in the node set. Note that the first node may be
   * random as this is a HashSet.
   *
   * @return The first node.
   */
  def getFirstNode(): N =
  {
    return nodes.head
  }

  override def toString: String =
  {
    return getClass.getSimpleName + "[" + hashCode + ", Nodes: " + nodes.size + "]"
  }
}